Implement produce actions in Python

bjustin-ibm opened the pull request:
Supersedes #127

bjustin-ibm commented:
@dubeejw Please review. However, I'm about to juggle the commits around a little, but theoretically none of the code will change as a result.

bjustin-ibm commented:
Depends on https://github.com/openwhisk/openwhisk/pull/1829

dubeejw commented on installCatalog.sh:
```diff
@@ -56,3 +56,6 @@ $WSK_CLI -i --apihost "$EDGEHOST" action update messaging/messageHubFeed "$PACKA
     -a parameters '[ {"name":"kafka_brokers_sasl", "required":true, "description": "Array of Message Hub brokers"},{"name":"user", "required":true, "description": "Message Hub username"},{"name":"password", "required":true, "description": "Message Hub password", "type":"password"},{"name":"topic", "required":true, "description": "Topic to subscribe to"},{"name":"isJSONData", "required":false, "de ...
     -a sampleInput '{"kafka_brokers_sasl":"[\"kafka01-prod01.messagehub.services.us-south.bluemix.net:9093\"]", "username":"someUsername", "password":"somePassword", "topic":"mytopic", "isJSONData": "false", "endpoint":"openwhisk.ng.bluemix.net", "kafka_admin_url":"https://kafka-admin-prod01.messagehub.services.us-south.bluemix.net:443"}'
+
+$WSK_CLI -i --apihost "$EDGEHOST" action update messaging/messageHubProduce "$PACKAGE_HOME/action/messageHubProduce.py" \
```
Needs annotations for description, parameters, and sampleInput.

bjustin-ibm replied:
I suppose README.md should be updated as well.

dubeejw commented on installKafka.sh:
```diff
@@ -47,3 +47,6 @@ $WSK_CLI -i --apihost "$EDGEHOST" action update messaging/kafkaFeed "$PACKAGE_HO
     -a parameters '[ {"name":"brokers", "required":true, "description": "Array of Kafka brokers"}, {"name":"topic", "required":true, "description": "Topic to subscribe to"}, {"name":"isJSONData", "required":false, "description": "Attempt to parse message content as JSON"}, {"name":"endpoint", "required":true, "description": "Hostname and port of OpenWhisk deployment"}]' \
     -a sampleInput '{"brokers":"[\"127.0.0.1:9093\"]", "topic":"mytopic", "isJSONData":"false", "endpoint": "openwhisk.ng.bluemix.net"}'
+
+$WSK_CLI -i --apihost "$EDGEHOST" action update messaging/kafkaProduce "$PACKAGE_HOME/action/kafkaProduce.py" \
```
Needs annotations for description, parameters, and sampleInput.

dubeejw commented on tests/src/system/packages/MessageHubProduceTests.scala:
```diff
@@ -0,0 +1,119 @@
+        "password" -> kafkaUtils("password").asInstanceOf[String].toJson,
+        "topic" -> topic.toJson,
+        "kafka_brokers_sasl" -> kafkaUtils("brokers").asInstanceOf[List[String]].toJson,
+        "value" -> "Big Trouble is actually a really good Tim Allen movie. Seriously.".toJson)
```
Shakes head.

bjustin-ibm replied:
Don't be a Gators fan...

dubeejw commented on action/messageHubProduce.py:
```diff
@@ -0,0 +1,55 @@
+    return {"success": True}
+
+def validateParams(params):
+    requiredParams = ['kafka_brokers_sasl', 'user', 'password', 'topic', 'value']
```
Is the `value` parameter actually the message being posted to the topic? If so, maybe it would be clearer as to what the parameter was if we called it `message`.

bjustin-ibm replied:
Yeah, "value" seems to be the Kafka terminology for this. Actually, the "message" (at least from the consumer side) includes lots of fields (key, partition, topic, etc.) and not just the value.

dubeejw commented on tests/src/system/packages/MessageHubProduceTests.scala:
```diff
@@ -0,0 +1,119 @@
+
+    // these parameter values are 100% valid and should work as-is
+    val validParameters = Map(
+        "user" -> kafkaUtils("user").asInstanceOf[String].toJson,
```
Why not have the apply() method in KafkaUtils return JSON instead of repeating `asInstanceOf[String].toJson` every time the method function is called?

bjustin-ibm replied:
I'll look into this. I think it was needed to not be JSON when we were still creating a separate KafkaProducer to produce a message during the test. Since this is gone, seems like we can switch this to always return JSON instead.

dubeejw commented on tests/src/system/utils/KafkaUtils.scala:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    def apply(key : String) = {
+        this.messageHubProps.getOrElse(key, "")
```
Return JSON representation of the field value.

bjustin-ibm replied:
I implemented `KafkaUtils.getAsJson()` instead.

bjustin-ibm commented:
@dubeejw Alright, I think I addressed all your concerns. Please review but do not merge (we're still waiting for https://github.com/openwhisk/openwhisk/pull/1829

csantanapr commented:
Closes #18
